src/manifolds/Tucker.jl

```diff
@@ -0,0 +1,812 @@
+    ξ = convert(Vector{promote_type(number_eltype(𝔄), eltype(x))}, x)
+    ℭ = 𝔄.hosvd.core
+    σ = 𝔄.hosvd.σ
+    U⊥ = ℬ.data.U⊥
+    D = ndims(ℭ)
+    r⃗ = size(ℭ)
+    n⃗ = size(𝔄)
+
+    # split ξ into ξ_core and ξU so that vcat(ξ_core, ξU...) == ξ, but avoid copying
+    ξ_core = view(ξ, 1:length(ℭ))
+    ξU = Vector{typeof(ξ_core)}(undef, D)
+    nextcolumn = length(ℭ) + 1
+    for d in 1:D
+        numcols = r⃗[d] * (n⃗[d] - r⃗[d])
+        ξU[d] = view(ξ, nextcolumn:(nextcolumn + numcols - 1))
+        nextcolumn += numcols
+    end
+
+    # Construct ∂U[d] by plugging in the definition of the orthonormal basis [Dewaele2021]
+    # ∂U[d] = ∑ᵢⱼ { ξU[d]ᵢⱼ (σ[d]ⱼ)⁻¹ U⊥[d] 𝐞ᵢ 𝐞ⱼᵀ }
+    #       = U⊥[d] * ∑ⱼ (σ[d]ⱼ)⁻¹ (∑ᵢ ξU[d]ᵢⱼ  𝐞ᵢ) 𝐞ⱼᵀ
+    # ξU[d] = [ξ₁₁, ..., ξ₁ⱼ, ..., ξᵢ₁, ..., ξᵢⱼ, ..., ]
+    # => turn these i and j into matrix indices and do matrix operations
+    for d in 1:D
+        grid = transpose(reshape(ξU[d], r⃗[d], n⃗[d] - r⃗[d]))
+        y.U̇[d] .= U⊥[d] * grid * Diagonal(1 ./ σ[d])
+    end
+
+    y.Ċ .= reshape(ξ_core, size(y.Ċ))
+    return y
+end
+function get_vector!(ℳ::Tucker, y, 𝔄::TuckerPoint, x, ℬ::DefaultOrthonormalBasis)
+    return get_vector!(ℳ, y, 𝔄, x, get_basis(ℳ, 𝔄, ℬ))
+end
+
+function get_vectors(ℳ::Tucker, 𝔄::TuckerPoint{T,D}, ℬ::CachedHOSVDBasis) where {T,D}
+    vectors = Vector{TuckerTVector{T,D}}(undef, manifold_dimension(ℳ))
+    foreach((i, vᵢ) -> setindex!(vectors, copy(vᵢ), i), ℳ, 𝔄, ℬ)
+    return vectors
+end
+function get_vectors(ℳ::Tucker, 𝔄::TuckerPoint, ℬ::DefaultOrthonormalBasis)
+    return get_vectors(ℳ, 𝔄, get_basis(ℳ, 𝔄, ℬ))
+end
+
+"""
+    inner(::Tucker, A::TuckerPoint, x::TuckerTVector, y::TuckerTVector)
+
+The Euclidean inner product between tangent vectors `x` and `y` at the point `A` on
+the Tucker manifold.function
+
+    inner(::Tucker, A::TuckerPoint, x::TuckerTVector, y)
+    inner(::Tucker, A::TuckerPoint, x, y::TuckerTVector)
+
+The Euclidean inner product between `x` and `y` where `x` is a vector tangent to the Tucker
+manifold at `A` and `y` is a vector or the ambient space or vice versa. The vector in the
+ambient space is represented as a full tensor.
+"""
+function inner(::Tucker, 𝔄::TuckerPoint, x::TuckerTVector, y::TuckerTVector)
+    ℭ = 𝔄.hosvd.core
+    dotprod = dot(x.Ċ, y.Ċ)
+    for d in 1:ndims(𝔄)
+        dotprod += dot(x.U̇[d] * tensor_unfold(ℭ, d), y.U̇[d] * tensor_unfold(ℭ, d))
+    end
+    return dotprod
+end
+inner(M::Tucker, 𝔄::TuckerPoint, x::TuckerTVector, y) = dot(embed(M, 𝔄, x), y)
+inner(M::Tucker, 𝔄::TuckerPoint, x, y::TuckerTVector) = dot(x, embed(M, 𝔄, y))
+
+"""
+    inverse_retract(ℳ::Tucker, A::TuckerPoint, B::TuckerPoint, r::ProjectionInverseRetraction)
+
+The projection inverse retraction on the Tucker manifold interprets `B` as a point in the
+ambient Euclidean space and projects it onto the tangent space at to `ℳ` at `A`.
+"""
+inverse_retract(
+    ::Tucker, ::Any, ::TuckerPoint, ::TuckerPoint, ::ProjectionInverseRetraction
```

Comment:
```suggestion
    ::Tucker, ::TuckerPoint, ::TuckerPoint, ::ProjectionInverseRetraction
```
Note that `inverse_retraction` as correctly documented has a manifold, two points and a retraction type, the any is a parameter too much here.